### PR TITLE
feat(core): implement support for Stateless MCP (SEP-2575)

### PR DESCRIPTION
**Pull request by anubhav756**
## Overview
Introduces reference support for the Stateless MCP spec draft (`DRAFT-2026-v1`) in alignment with SEP-2575. Adds a new `v20260618` transport layer supporting HTTP header-based request metadata, protocol version headers, client/server metadata propagation (`_meta`), dynamic protocol negotiation fallback, and expanded parameter schema validation for numeric types.

The `DRAFT-2026-v1` specification introduces a stateless execution model for MCP over HTTP. Rather than requiring stateful connection handshakes, requests carry routing and context via dedicated HTTP headers (`Mcp-Method`, `Mcp-Name`, `MCP-Protocol-Version`) and embedded `_meta` fields in parameters, optimizing performance and server scalability.

## Changes
- **Core Protocol Registration**:
  - Added `MCPv20260618` (`"DRAFT-2026-v1"`) protocol constant and `MCPDraft` alias.
  - Integrated `v20260618` transport instantiation inside `NewToolboxClient`.
  - Defined request/response structures for `DRAFT-2026-v1`, including `MCPMeta` and `MCPResultMeta`.
  - Implemented `McpTransport` with support for HTTP headers (`MCP-Protocol-Version`, `Mcp-Method`, `Mcp-Name`).
  - Added unit tests for tool discovery/call, header propagation, schema parsing, and protocol fallback error codes (`-32022`, `-32004`).
  - Added `ProtocolNegotiationError` for fallback signaling.
  - Updated `ValidateType` and `schemaToMap` to support JSON Schema `"number"` types (`float32`, `float64`, `int`, etc.).

### CODE REVIEW

File: core/protocol.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ const (
 	// MCP is the default alias pointing to the newest supported version.
 	MCP = MCPv20250618
 
-	MCPLatest = MCPv20251125
```
Comment (dishaprakash) [inline core/protocol.go:32]: Is there a reason we are downgrading this?

File: core/client.go [hunk 2]
```diff
@@ -251,6 +256,17 @@ func (tc *ToolboxClient) LoadTool(name string, ctx context.Context, opts ...Tool
...
+	var negErr *ProtocolNegotiationError
+	if err != nil && errors.As(err, &negErr) {
+		newTransport, transErr := tc.createTransport(Protocol(negErr.FallbackVersion))
+		if transErr != nil {
+			return nil, fmt.Errorf("failed to fallback to protocol %s: %w", negErr.FallbackVersion, transErr)
+		}
+		tc.transport = newTransport
+		tc.protocol = Protocol(negErr.FallbackVersion)
+		manifest, err = tc.transport.GetTool(ctx, name, resolvedHeaders)
+	}
```
Comment (dishaprakash) [inline core/client.go]: Can we add tests to:
1. verify the fallback is successful
2. Edge cases (maybe in the case a goroutine is being used, to ensure thread safety)

File: core/transport/mcp/v20250618/mcp.go [hunk 3]
```diff
@@ -253,6 +253,23 @@ func (t *McpTransport) doRPC(ctx context.Context, url string, reqBody any, heade
 		// Continue to body parsing
 	} else if (resp.StatusCode == http.StatusAccepted || resp.StatusCode == http.StatusNoContent) && dest == nil {
 		return nil // Valid notification success
+	} else if resp.StatusCode == http.StatusBadRequest {
+		body, _ := io.ReadAll(resp.Body)
```
Comment (dishaprakash) [inline core/transport/mcp/v20250618/mcp.go]: Should we de-duplicate this code and add it as a helper in the base mcp class?

File: core/protocol.go (hunk 1 above)
Comment (anubhav756) [reply core/protocol.go:32]: This is updated to point to the 2026 version but this would change in a following PR since we would have MCPDraft to point to that. Please note that this would again change once we hit MCP's official dated release. Then we would move MCPLatest to point to the 2026 version only, and MCPDraft would be updated whenever MCP comes up with another draft release.

Update: I pulled that change back so now it still points to 2025.

File: core/client.go (hunk 2 above)
Comment (anubhav756) [reply core/client.go]: Thanks for pointing those out. Added them in the future PR (to avoid conflicts): https://github.com/googleapis/mcp-toolbox-sdk-go/pull/306

File: core/transport/mcp/v20250618/mcp.go (hunk 3 above)
Comment (anubhav756) [reply core/transport/mcp/v20250618/mcp.go]: Thanks for bringing this up. We try to make the different transports as independent as possible, just like in python. However, I did dedupe some other part of code anyways. Thanks! :)

File: core/transport/mcp/v20260618/mcp.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to build toolset URL: %w", err)
+		}
+	} else {
+		targetURL = t.BaseURL()
+	}
+
+	reqPayload := ListToolsRequest{
+		JSONRPC: "2.0",
+		ID:      "1",
```
Comment (Yuan325) [inline core/transport/mcp/v20260618/mcp.go]: Should we hard code the id here? Can you refer back to the previous version protocol?

Also, should we build the jsonRPCRequest in the sendRequest function instead of here? (similar to the previous impleemntation).

File: core/transport/mcp/v20260618/mcp.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	toolsMap := make(map[string]transport.ToolSchema)
+	for _, toolRaw := range result.Tools {
+		toolSchema, err := t.ConvertToolDefinition(toolRaw)
+		if err != nil {
+			continue
+		}
+		if name, ok := toolRaw["name"].(string); ok {
+			toolsMap[name] = toolSchema
+		}
+	}
```
Comment (Yuan325) [inline core/transport/mcp/v20260618/mcp.go]: I don't think we update the toolsMap from the previous version? Can you keep it as is? We might need to update the type including the following: 
1. result meta that consist of server information
2. `ttlMs` and `cacheScope` field
3. authparam vendor prefix that was updated to `com.google.cloud`.


I don't think we need to update any parsing logics here.

File: core/transport/mcp/v20260618/mcp.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+	return &transport.ManifestSchema{
+		ServerVersion: "1.0.0",
```
Comment (Yuan325) [inline core/transport/mcp/v20260618/mcp.go]: This shouldn't be hard coded

File: core/transport/mcp/v20260618/mcp.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+		Tools:         toolsMap,
+	}, nil
+}
+
+// InvokeTool executes a tool on the server.
+func (t *McpTransport) InvokeTool(ctx context.Context, toolName string, payload map[string]any, headers map[string]string) (any, error) {
```
Comment (Yuan325) [inline core/transport/mcp/v20260618/mcp.go:129]: Similarly can we keep the implementation similar to the previous version's? And only update the part that is updated.

**Review by Yuan325 — CHANGES_REQUESTED** [review]
Can we refactor this PR? Most code should be similar with previous implementation. With only updating the parts that are associated with the SEP-2575 change: 

1. Removing initialize
2. Adding server discover
3. Adding the metadata in every request sent (which I believe could just be added in the `sendRequest()` function
4. Reading the serverInfo from `_meta` result (I believe we only needed to update the `types.go`).

File: core/transport/mcp/v20260618/mcp.go (hunk 6 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go]: This is updated dynamically in the future PR https://github.com/googleapis/mcp-toolbox-sdk-go/pull/308

**Comment by anubhav756** [comment]
> Can we refactor this PR? Most code should be similar with previous implementation. With only updating the parts that are associated with the SEP-2575 change:
> 
> 1. Removing initialize
> 2. Adding server discover
> 3. Adding the metadata in every request sent (which I believe could just be added in the `sendRequest()` function
> 4. Reading the serverInfo from `_meta` result (I believe we only needed to update the `types.go`).

This is covered in https://github.com/googleapis/mcp-toolbox-sdk-go/pull/308.

File: core/transport/mcp/v20260618/mcp.go (hunk 4 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go]: Done in https://github.com/googleapis/mcp-toolbox-sdk-go/pull/309

File: core/transport/mcp/v20260618/mcp.go (hunk 5 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go]: Covered in https://github.com/googleapis/mcp-toolbox-sdk-go/pull/309. Thanks!

File: core/transport/mcp/v20260618/mcp.go (hunk 7 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go:129]: Addressed in https://github.com/googleapis/mcp-toolbox-sdk-go/pull/309

**Review by Yuan325 — APPROVED** [review]

**Comment by Yuan325** [comment]
I'll approve this for now. Seems like the SEP-2575 changes were implemented across multiple PRs. Please double check the implementation before merging the `mcp-v202606` branch to make sure the hardcoded informations are not merged in. Tyyy

File: core/transport/mcp/v20260618/mcp.go (hunk 5 above)
Comment (Yuan325) [reply core/transport/mcp/v20260618/mcp.go]: Seems like this wasn't reflected in 309, please check~

File: core/transport/mcp/v20260618/mcp.go (hunk 7 above)
Comment (Yuan325) [reply core/transport/mcp/v20260618/mcp.go:129]: Seems like this wasn't reflected. Please check.

File: core/transport/mcp/v20260618/mcp.go (hunk 5 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go]: I think I should have clarified better:

* https://github.com/googleapis/mcp-toolbox-sdk-go/pull/309: updates the type
* https://github.com/googleapis/mcp-toolbox-sdk-go/pull/308: updates the result meta for server information
* https://github.com/googleapis/mcp-toolbox-sdk-go/pull/315: addresses the vendor prefix, `ttlMs` and `cacheScope` fields.

> [!NOTE]
> I had to spread these in order to minimize conflicts. Nevertheless, I will double-check these in the base branch `mcp-v202606` once staged. Thanks for bringing these up though :)

File: core/transport/mcp/v20260618/mcp.go (hunk 7 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go:129]: I assume these parts from #309 address the concerns: https://screenshot-v2.corp.google.com/AeVxsUVdt4G3xQW

I know it's a bit tricky to review since it's spread across multiple PRs since I had optimized for small PRs as much as possible but also tried to avoid too many conflicts.

Nevertheless, I will double-check this in the staging PR as well. Thanks for brining this up!

PS: Here's the final staging PR: https://github.com/googleapis/mcp-toolbox-sdk-go/pull/317

File: core/transport/mcp/v20260618/mcp.go (hunk 5 above)
Comment (anubhav756) [reply core/transport/mcp/v20260618/mcp.go]: Quick update: I've cleared out ttlMs and cacheScope related changes to do that in a followup PR cleanly, according to https://docs.google.com/document/d/1E-iNN2vL4CROPJdSF4wWvyYvH2vgIhNQKQ3fPowOnyY/edit?tab=t.0
